linera-views/src/backends/dynamo_db.rs

```diff
@@ -697,73 +676,40 @@ struct QueryResponses {
     responses: Vec<QueryOutput>,
 }
 
-// Inspired by https://depth-first.com/articles/2020/06/22/returning-rust-iterators/
-#[doc(hidden)]
-#[expect(clippy::type_complexity)]
-pub struct DynamoDbKeyBlockIterator<'a> {
-    prefix_len: usize,
-    pos: usize,
-    iters: Vec<
-        std::iter::Flatten<
-            std::option::Iter<'a, Vec<HashMap<std::string::String, AttributeValue>>>,
-        >,
-    >,
-}
-
-impl<'a> Iterator for DynamoDbKeyBlockIterator<'a> {
-    type Item = Result<&'a [u8], DynamoDbStoreInternalError>;
-
-    fn next(&mut self) -> Option<Self::Item> {
-        let result = self.iters[self.pos].next();
-        match result {
-            None => {
-                if self.pos == self.iters.len() - 1 {
-                    return None;
-                }
-                self.pos += 1;
-                self.iters[self.pos]
-                    .next()
-                    .map(|x| extract_key(self.prefix_len, x))
-            }
-            Some(result) => Some(extract_key(self.prefix_len, result)),
+impl QueryResponses {
+    fn iterate_keys(&self) -> DynamoDbKeyBlockIterator<'_> {
+        let pos = 0;
+        let mut iters = Vec::new();
+        for response in &self.responses {
+            let iter = response.items.iter().flatten();
+            iters.push(iter);
+        }
+        DynamoDbKeyBlockIterator {
+            prefix_len: self.prefix_len,
+            pos,
+            iters,
         }
     }
-}
 
-/// A set of keys returned by a search query on DynamoDB.
-pub struct DynamoDbKeys {
-    result_queries: QueryResponses,
-}
-
-impl KeyIterable<DynamoDbStoreInternalError> for DynamoDbKeys {
-    type Iterator<'a>
-        = DynamoDbKeyBlockIterator<'a>
-    where
-        Self: 'a;
-
-    fn iterator(&self) -> Self::Iterator<'_> {
+    fn iterate_key_values(&self) -> DynamoDbKeyValueIterator<'_> {
         let pos = 0;
         let mut iters = Vec::new();
-        for response in &self.result_queries.responses {
+        for response in &self.responses {
             let iter = response.items.iter().flatten();
             iters.push(iter);
         }
-        DynamoDbKeyBlockIterator {
-            prefix_len: self.result_queries.prefix_len,
+        DynamoDbKeyValueIterator {
+            prefix_len: self.prefix_len,
             pos,
             iters,
         }
     }
 }
 
-/// A set of `(key, value)` returned by a search query on DynamoDB.
-pub struct DynamoDbKeyValues {
-    result_queries: QueryResponses,
-}
-
+// Inspired by https://depth-first.com/articles/2020/06/22/returning-rust-iterators/
```

Comment:
Why not use `impl Iterator`? I don't think either of the arguments against it apply here.

```rust
self.responses
    .iter()
    .flat_map(|response| &response.items)
    .map(move |item| extract_key(self.prefix_len, item))
```

(Edit: Ah, you say it would be slightly slower? But I don't see why.)